File: python/timeit.py
import sys, time

_range = xrange if sys.version_info.major == 2 else range
_timefn = time.time if not sys.platform == 'win32' else time.clock

class Timeit(object):
    
    def __init__(self):
        self.average = 0.0
        self.mean = 0.0
        self.min = 0.0
        self.max = 0.0
        
    def run(self, count, func, *args):
        
        times = []
        result = None
        for i in _range(count):
            tstart = _timefn()
            
            result = func(*args)
            
            tend = _timefn()
            times.append( tend - tstart )
        
        times.sort()
        self.average = sum( times ) / float(count)
        middle = count//2
        if count & 1:
            self.mean = times[middle]
        else:
            self.mean = (times[middle-0] + times[middle]) / 2.0
        self.min = times[0]
        self.max = times[-1]
        
        return result
            
            

File: cs/timeit.cs
using System;
using System.Linq;
using System.Diagnostics;
using System.Collections.Generic;

public class Timeit
{
    public double Average;
    public double Mean;
    public double Min;
    public double Max;

    public TResult Run<TResult>(ulong count, Func<TResult> func, params object[] args) {
        return RunInternal<TResult>(count, func, args);
    }
    public TResult Run<T1,TResult>(ulong count, Func<T1,TResult> func, params object[] args) {
        return RunInternal<TResult>(count, func, args);
    }
    public TResult Run<T1,T2,TResult>(ulong count, Func<T1,T2,TResult> func, params object[] args) {
        return RunInternal<TResult>(count, func, args);
    }

    private TResult RunInternal<TResult>(ulong count, Delegate func, params object[] args)
    {
        TimeSpan[] times = new TimeSpan[count];
        
        TResult result = default(TResult);
        for(ulong i = 0; i < count; ++i)
        {
            Stopwatch watch = new Stopwatch();
            watch.Start();
            
            result = (TResult)func.DynamicInvoke(args);
            
            watch.Stop();
            
            times[i] = watch.Elapsed;
        }
        
        Array.Sort(times);
        long averageTicks = Convert.ToInt64(times.Average(timeSpan => timeSpan.Ticks));
        TimeSpan average = new TimeSpan(averageTicks);
        Average = average.TotalSeconds;
        
        ulong middle = count / 2;
        if( (count & 1) == 1 )
            Mean = times[middle].TotalSeconds;
        else
            Mean = new TimeSpan((times[middle-1].Ticks + times[middle].Ticks) / 2).TotalSeconds;
        
        Min = times[0].TotalSeconds;
        Max = times[count-1].TotalSeconds;
        
        return result;
    }
    

}